zingo-sync/src/scan.rs

```diff
@@ -0,0 +1,322 @@
+    consensus::{BlockHeight, NetworkUpgrade, Parameters},
+    transaction::TxId,
+    zip32::AccountId,
+};
+
+use crate::{
+    client::{self, get_compact_block_range, FetchRequest},
+    primitives::{OutputId, WalletCompactBlock},
+};
+
+use self::runners::BatchRunners;
+
+pub(crate) mod runners;
+
+type KeyId = (AccountId, Scope);
+
+struct InitialScanData<'a> {
+    previous_block: Option<&'a WalletCompactBlock>,
+    sapling_initial_tree_size: u32,
+    orchard_initial_tree_size: u32,
+}
+
+impl InitialScanData<'_> {
+    async fn new<P>(
+        fetch_request_sender: UnboundedSender<FetchRequest>,
+        parameters: &P,
+        first_block: &CompactBlock,
+    ) -> Result<Self, ()>
+    where
+        P: Parameters + Send + 'static,
+    {
+        // TODO: get last block of adjacent lower scan range a.k.a. previous_block
+        let previous_block: Option<&WalletCompactBlock> = None;
+
+        // gets initial tree size from previous block if available
+        // otherwise, from first block if available
+        // otherwise, fetches frontiers from server
+        let (sapling_initial_tree_size, orchard_initial_tree_size) =
+            if let Some(prev) = previous_block {
+                (
+                    prev.sapling_commitment_tree_size(),
+                    prev.orchard_commitment_tree_size(),
+                )
+            } else if let Some(chain_metadata) = first_block.chain_metadata.clone() {
+                // calculate initial tree size by subtracting number of outputs in block from the blocks final tree size
+                let sapling_output_count: u32 = first_block
+                    .vtx
+                    .iter()
+                    .map(|tx| tx.outputs.len())
+                    .sum::<usize>()
+                    .try_into()
+                    .expect("Sapling output count cannot exceed a u32");
+                let orchard_output_count: u32 = first_block
+                    .vtx
+                    .iter()
+                    .map(|tx| tx.actions.len())
+                    .sum::<usize>()
+                    .try_into()
+                    .expect("Sapling output count cannot exceed a u32");
+
+                (
+                    chain_metadata
+                        .sapling_commitment_tree_size
+                        .checked_sub(sapling_output_count)
+                        .unwrap(),
+                    chain_metadata
+                        .orchard_commitment_tree_size
+                        .checked_sub(orchard_output_count)
+                        .unwrap(),
+                )
+            } else {
+                let sapling_activation_height = parameters
+                    .activation_height(NetworkUpgrade::Sapling)
+                    .expect("should have some sapling activation height");
+
+                match first_block.height().cmp(&sapling_activation_height) {
+                    Ordering::Greater => {
+                        let frontiers =
```

Comment:
so we go to the trouble of downloading the frontiers, ought we to do something with them?

Reply:
yeah thats line 99 and 100